cryptoservice/crypto_service.go

```diff
@@ -105,13 +99,33 @@ func (cs *CryptoService) GetKey(keyID string) data.PublicKey {
 	return data.PublicKeyFromPrivate(privKey)
 }
 
+// GetKeyInfo returns role and GUN info of a key by ID
+func (cs *CryptoService) GetKeyInfo(keyID string) (trustmanager.KeyInfo, error) {
+	for _, store := range cs.keyStores {
+		if info, err := store.GetKeyInfo(keyID); err == nil {
+			return info, nil
+		}
+	}
+	return trustmanager.KeyInfo{}, fmt.Errorf("Could not find info for keyID %s", keyID)
+}
+
 // RemoveKey deletes a key by ID
 func (cs *CryptoService) RemoveKey(keyID string) (err error) {
-	keyPaths := []string{keyID, filepath.Join(cs.gun, keyID)}
 	for _, ks := range cs.keyStores {
-		for _, keyPath := range keyPaths {
-			ks.RemoveKey(keyPath)
-		}
+		ks.RemoveKey(keyID)
+	}
+	return // returns whatever the final values were
+}
+
+// AddKey adds a private key to a specified role.
+// The GUN is inferred from the cryptoservice itself for non-root roles
+func (cs *CryptoService) AddKey(role string, key data.PrivateKey) (err error) {
+	keyID := key.ID()
+	if role != data.CanonicalRootRole {
+		keyID = filepath.Join(cs.gun, key.ID())
+	}
+	for _, ks := range cs.keyStores {
```

Comment:
You were right: I was continuing to add to other keystores earlier, but I think bailing is the right behavior.  Done